### PR TITLE
 feat(aio): grouped search results

### PR DESCRIPTION
**Pull request by petebacondarwin**
The basic functionality is ready. The main feature missing is to hide the results when you click outside the results list. That is provided in #15096

### CODE REVIEW

File: aio/src/app/search/search-results/search-results.component.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
+import { Component, ChangeDetectionStrategy, EventEmitter, HostListener, Input, OnInit, Output } from '@angular/core';
```
Comment (gkalpak) [inline aio/src/app/search/search-results/search-results.component.ts]: `Input` not used.

File: aio/src/app/search/search-results/search-results.component.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import { SearchResult, SearchResults, SearchService } from '../search.service';
+
+export interface SearchArea {
+  name: string;
+  pages: SearchResult[];
+}
+
+/**
+ * A component to display the a list of search results
```
Comment (gkalpak) [inline aio/src/app/search/search-results/search-results.component.ts]: _the a_ --> _the_ or _a_ :smiley:

File: aio/src/app/search/search-results/search-results.component.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  showResults = false;
+
+  @Output()
+  resultSelected = new EventEmitter<SearchResult>();
+
+  /**
+   * A mapping of the search results grouped into areas
+   */
+  searchAreas = new ReplaySubject<SearchArea[]>();
```
Comment (gkalpak) [inline aio/src/app/search/search-results/search-results.component.ts]: Do we need to replay anything but the last value?

File: aio/src/app/search/search-results/search-results.component.spec.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        { path: 'news', title: 'News', type: 'marketing', keywords: '', titleWords: '' }
+      ] }
+    ]);
+  });
+
+  it('should emit an "resultSelected" event when a search result anchor is clicked', () => {
+    let selectedResult: SearchResult;
+    component.resultSelected.subscribe((result: SearchResult) => selectedResult = result);
+    const results = [
+      {path: 'news', title: 'News', type: 'marketing', keywords: '', titleWords: '' }
```
Comment (gkalpak) [inline aio/src/app/search/search-results/search-results.component.spec.ts]: Inconsistent whitespace here and below: `{page...'' }` 😱 😱 😱

File: aio/src/app/search/search-box/search-box.component.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      this.onSearch(query);
+    }
+  }
+
+  onSearch(query: string, keyCode?: number) {
+    if (keyCode === 27) {
+      // Ignore escape key
+      return;
+    }
+    this.locationService.setSearch('Full Text Search', { search: query });
```
Comment (gkalpak) [inline aio/src/app/search/search-box/search-box.component.ts]: I just noticed that the first argument to `setSearch` [is ignored][1]. Is it a bug? (Not related to this PR - just verifying.)

[1]: https://github.com/petebacondarwin/angular/blob/5d1de0789a668f9c596116b5207071601212a7ec/aio/src/app/shared/location.service.ts#L60

File: aio/src/app/search/search-box/search-box.component.spec.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    input.triggerEventHandler('keyup', { target: { value: 'some query'}});
+    expect(search.search).toHaveBeenCalledWith('some query');
+  }));
+
+  it('should call the search service on focus', inject([SearchService], (search: SearchService) => {
+    const input = fixture.debugElement.query(By.css('input'));
+    input.triggerEventHandler('focus', { target: { value: 'some query'}});
+    expect(search.search).toHaveBeenCalledWith('some query');
+  }));
+});
```
Comment (gkalpak) [inline aio/src/app/search/search-box/search-box.component.spec.ts]: Missing a couple of tests:

- Calling `LocationService#setSearch()`.
- Ignoring the <kbd>ESC</kbd> key.
- Initializing based on URL query.

**Comment by gkalpak** [comment]
BTW, if the search is costly, we could avoid redoing it when focusing the search-box. We could cache the last result and return that if the query matches. But doesn't have to be part of this PR (if at all necessary in the first place).

File: aio/src/app/search/search-box/search-box.component.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+ */
+@Component({
+  selector: 'aio-search-box',
+  templateUrl: './search-box.component.html',
+  styleUrls: ['./search-box.component.scss']
+})
+export class SearchBoxComponent implements OnInit {
+
+  @ViewChild('searchBox') searchBox: ElementRef;
+  constructor(private searchService: SearchService, private locationService: LocationService) { }
```
Comment (petebacondarwin) [inline aio/src/app/search/search-box/search-box.component.ts]: done

File: aio/src/app/search/search-box/search-box.component.ts (hunk 5 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-box/search-box.component.ts]: It is a bug in that service, which is fixed in another of my PRs :-)

https://github.com/angular/angular/pull/15090/files#diff-d7053af3a910e9defc92631371cb06efR59

File: aio/src/app/search/search-results/search-results.component.ts (hunk 1 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-results/search-results.component.ts]: fixed

File: aio/src/app/search/search-results/search-results.component.ts (hunk 2 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-results/search-results.component.ts]: fixed

File: aio/src/app/search/search-results/search-results.component.ts (hunk 3 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-results/search-results.component.ts]: only one - good spot.

File: aio/src/app/search/search-results/search-results.component.spec.ts (hunk 4 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-results/search-results.component.spec.ts]: fixed

File: aio/src/app/search/search-box/search-box.component.spec.ts (hunk 6 above)
Comment (petebacondarwin) [inline aio/src/app/search/search-box/search-box.component.spec.ts]: added

**Comment by petebacondarwin** [comment]
The search is done in a webworker so it is not costly from the user's perspective.
I agree that we could optimize it since we have the data available but I felt that it added unnecessary complexity for now practical gain.

File: aio/src/app/search/search-box/search-box.component.ts [hunk 8]
```diff
@@ -22,10 +22,12 @@ import { LocationService } from 'app/shared/location.service';
 export class SearchBoxComponent implements OnInit {
 
   @ViewChild('searchBox') searchBox: ElementRef;
+
   constructor(private searchService: SearchService, private locationService: LocationService) { }
 
   ngOnInit() {
     const query = this.locationService.search()['search'];
+    console.log('here', query, this.locationService.search());
```
Comment (petebacondarwin) [inline aio/src/app/search/search-box/search-box.component.ts]: doh!

File: aio/src/app/search/search-results/search-results.component.spec.ts (hunk 4 above)
Comment (gkalpak) [inline aio/src/app/search/search-results/search-results.component.spec.ts]: I don't think so 😛, but it shouldn't block this PR. I am pretty sure there are other places with inconsistent whitespace.

**Review by gkalpak — APPROVED** [review]

**Pull request merged**